plugins/system/languagefilter/src/Extension/LanguageFilter.php

```diff
@@ -327,171 +327,115 @@ public function postprocessNonSEFBuildRule(&$router, &$uri)
+                || \count($this->getApplication()->getInput()->post) > 0
+                || \count($this->getApplication()->getInput()->files) > 0
+            ) {
                 $lang_code = $this->default_lang;
-            }
-        }
-
-        // We have not found the language and thus need to redirect
-        if (!$found) {
-            // Lets find the default language for this user
-            if (!isset($lang_code) || !isset($this->lang_codes[$lang_code])) {
-                $lang_code = false;
+            } else {
+                $lang_code = $this->getLanguageCookie();
 
-                if ($this->params->get('detect_browser', 1)) {
+                // No language code. Try using browser settings or default site language
+                if (!$lang_code && $this->params->get('detect_browser', 0) == 1) {
                     $lang_code = LanguageHelper::detectLanguage();
-
-                    if (!isset($this->lang_codes[$lang_code])) {
-                        $lang_code = false;
-                    }
                 }
 
                 if (!$lang_code) {
                     $lang_code = $this->default_lang;
                 }
-            }
-
-            if ($this->mode_sef) {
-                // Use the current language sef or the default one.
-                if (
-                    $lang_code !== $this->default_lang
-                    || !$this->params->get('remove_default_prefix', 0)
-                ) {
-                    $path = $this->lang_codes[$lang_code]->sef . '/' . $path;
-                }
 
-                $uri->setPath($path);
-
-                if (!$this->getApplication()->get('sef_rewrite')) {
-                    $uri->setPath('index.php/' . $uri->getPath());
-                }
+                if (!$this->params->get('remove_default_prefix', 0) && $uri->getPath() == '') {
+                    if ($this->mode_sef) {
+                        $path = $this->lang_codes[$lang_code]->sef . '/' . $uri->getPath();
 
-                $redirectUri = $uri->base() . $uri->toString(['path', 'query', 'fragment']);
-            } else {
-                $uri->setVar('lang', $this->lang_codes[$lang_code]->sef);
-                $redirectUri = $uri->base() . 'index.php?' . $uri->getQuery();
-            }
+                        if (!$this->getApplication()->get('sef_rewrite')) {
+                            $path = 'index.php/' . $path;
+                        }
 
-            // Set redirect HTTP code to "302 Found".
-            $redirectHttpCode = 302;
+                        $uri->setPath($path);
+                    } else {
+                        $uri->setPath('index.php');
+                        $uri->setVar('lang', $this->lang_codes[$lang_code]->sef);
+                    }
+                    $redirectHttpCode = 301;
+                    $redirectUri = $uri->base() . $uri->toString(['path', 'query', 'fragment']);
```

Comment:
```suggestion
                    $redirectHttpCode = 301;
                    $redirectUri      = $uri->base() . $uri->toString(['path', 'query', 'fragment']);
```
php-cs-fixer suggestion 2 of 2